File: leetcode/python/322-coin-change.py
#!/usr/bin/env python
# encoding: utf-8

"""
You are given coins of different denominations and a total amount of money amount. Write a function to compute the fewest number of coins that you need to make up that amount. If that amount of money cannot be made up by any combination of the coins, return -1.

Example 1:
coins = [1, 2, 5], amount = 11
return 3 (11 = 5 + 5 + 1)

Example 2:
coins = [2], amount = 3
return -1.

Note:
You may assume that you have an infinite number of each kind of coin.
"""


class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        cnt = [-1 for i in range(amount + 1)]
        coins.sort().reverse()
        def find(mt):
            if mt < 0: return 0
            if mt == 0: return 1
            if cnt[mt] >= 0 : return cnt[mt]
            re = 0
            for c in coins:
                t = find(mt - c)
                if t > 0:
                    cnt[mt] = t + 1
                    return t + 1
            cnt[mt] = 0
            return re
        return find(amount)

File: leetcode/cpp/322-coin-change.cpp
/*
 * 计算使用的最少硬币数
 * 数组记录+ dfs
 *Example 1:
coins = [1, 2, 5], amount = 11
return 3 (11 = 5 + 5 + 1)

Example 2:
coins = [2], amount = 3
return -1.
 */



class Solution {
    vector<int> * cMin;
    vector<int> * co;
    int coLen;
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> t(1+amount, -2);
        cMin = &(t);
        co = &coins;
        coLen = coins.size();
        return getMin(amount);
    }
    int getMin(int n){
        if(n == 0) return 0;
        if(n < 0) return -1;
        if((*cMin)[n] > -2) return (*cMin)[n];
        int re = -1;
        for(int i = 0 ; i < coLen; i ++)
            if(n >= (*co)[i]){
                int cmin = getMin(n - (*co)[i]);
                if(cmin > -2)
                    re = re == -1 && cmin >= 0 ? cmin+1 : (cmin >= 0 ? min(re, cmin + 1) : re);
            }
        (*cMin)[n] = re;
        return re;
    }
};
